### PR TITLE
UPSTREAM: 49285: do not mutate statefulset on update

### PR DESCRIPTION
**Pull request by mfojtik**

### CODE REVIEW

**Comment by mfojtik** [comment]
[test]

File: vendor/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set.go [hunk 1]
```diff
@@ -450,7 +450,11 @@ func (ssc *StatefulSetController) sync(key string) error {
 // syncStatefulSet syncs a tuple of (statefulset, []*v1.Pod).
 func (ssc *StatefulSetController) syncStatefulSet(set *apps.StatefulSet, pods []*v1.Pod) error {
 	glog.V(4).Infof("Syncing StatefulSet %v/%v with %d pods", set.Namespace, set.Name, len(pods))
-	if err := ssc.control.UpdateStatefulSet(set, pods); err != nil {
+	setCopy, err := api.Scheme.DeepCopy(set)
```
Comment (mfojtik) [inline vendor/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set.go]: this should probably go deeper into UpdateStatefulSet when we know the mutation will happen, will tweak this if this confirms the tests pass.

Comment (deads2k) [inline vendor/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set.go]: > this should probably go deeper into UpdateStatefulSet when we know the mutation will happen, will tweak this if this confirms the tests pass.

It should go deeper, but if this works it's ok to bring out the big hammer to make progress as long as you still chase it down.

**Comment by mfojtik** [comment]
flake: https://github.com/openshift/origin/issues/15329

[test] again

File: vendor/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set.go (hunk 1 above)
Comment (liggitt) [inline vendor/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set.go]: I'm really surprised the mutation detector didn't catch this upstream...

Comment (liggitt) [inline vendor/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set.go]: I also didn't see anything in the call tree below this that mutated without copying first... I might be missing something, but the fact that the diff was only on the kind and apiVersion fields is odd to me

Comment (deads2k) [inline vendor/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set.go]: > I also didn't see anything in the call tree below this that mutated without copying first... I might be missing something, but the fact that the diff was only on the kind and apiVersion fields is odd to me

Suggests that it's getting converted, encoded, or something.  It wasn't obvious to me either.

**Comment by deads2k** [comment]
@mfojtik open upstream to match please.

**Comment by mfojtik** [comment]
@deads2k will give it one more [test] to be sure... do we have time to investigate what actually mutate or we are fine with this and follow up issue?

**Comment by deads2k** [comment]
> @deads2k will give it one more [test] to be sure... do we have time to investigate what actually mutate or we are fine with this and follow up issue?

I'd like to see an upstream PR opened, so that this title can be updated,so the future rebaser can figure out how to handle this.  Otherwise, no objection.

**Comment by mfojtik** [comment]
flake: https://github.com/openshift/origin/issues/15359

[test]

File: vendor/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set.go (hunk 1 above)
Comment (mfojtik) [inline vendor/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set.go]: yeah seems like decoded to internal version, I will open upstream pull and we can take it from there

Comment (mfojtik) [inline vendor/k8s.io/kubernetes/pkg/controller/statefulset/stateful_set.go]: does this looks suspicious? https://github.com/mfojtik/kubernetes/blob/35d6cf086c38f4416f72c91c8af138e8eb3e4090/pkg/controller/statefulset/stateful_set_utils.go#L303

**Comment by mfojtik** [comment]
@deads2k https://github.com/kubernetes/kubernetes/pull/49285

**Comment by mfojtik** [comment]
@deads2k i ran test on this 4 times and haven't seen the mutation shutdown

**Comment by deads2k** [comment]
[merge]

**Comment by deads2k** [comment]
Isolated change to fix a flake. Merging on green test.

**Pull request merged**